### ISSUE TITLE
[FEATURE] Feature Request: ViZDoom Environment Support

### ISSUE THREAD

**Issue by Old-Story** [Abdulhamid97Mousa/mosaic#31]
MOSAIC supports a wide range of environments but does not currently include **ViZDoom** (Kempka et al., IEEE CIG 2016), a Doom-based visual reinforcement learning platform maintained by the Farama Foundation. ViZDoom is the canonical benchmark for pixel-based 3D FPS reinforcement learning and has been used in landmark research on visual RL, curiosity-driven exploration, and hierarchical policies. Adding ViZDoom would give MOSAIC researchers a first-person 3D vision benchmark that complements the existing 2D pixel benchmarks (Atari ALE) and tensor-observation environments (SMAC, MultiGrid).

## What is ViZDoom?

  ViZDoom is a single-agent (and optionally multi-agent) visual RL platform built on the ZDoom engine, exposing:

  - **Pixel observations**: 320×240 RGB frames by default, configurable to arbitrary resolutions and to grayscale
  - **Discrete action space**: MOVE_FORWARD/BACKWARD, MOVE_LEFT/RIGHT, TURN_LEFT/RIGHT, ATTACK, SELECT_WEAPON_*, USE, RELOAD (buttons compose bitmask-style)
  - **Auxiliary sensors** (opt-in): depth buffer, semantic labels buffer, automap buffer, game variables (HP, ammo, weapon, position, kill count)
  - **8 curated scenarios** in order of difficulty:
    - `basic.cfg` (shoot a stationary target, 1-action sanity check)
    - `defend_the_center.cfg` (360° turret defense with limited ammo)
    - `defend_the_line.cfg` (linear defense against advancing enemies)
    - `health_gathering.cfg` (navigate slime maze collecting health packs)
    - `my_way_home.cfg` (maze navigation to a fixed goal, canonical curiosity/exploration test)
    - `deadly_corridor.cfg` (kill 6 enemies to reach a vest)
    - `predict_position.cfg` (lead a moving target with a rocket)
    - `take_cover.cfg` (dodge incoming fireballs)
  - **Multiplayer / bot support**: 2 to 8 agent deathmatch or cooperative modes via ZDoom's built-in networking
  - **Frame skip & sync modes**: `set_frames_per_action`, `set_ticrate`, sync/async modes for reproducibility vs. wall-clock play

  Repository: [Farama-Foundation/ViZDoom](https://github.com/Farama-Foundation/ViZDoom)
  Paper: Kempka et al., "ViZDoom: A Doom-based AI Research Platform for Visual Reinforcement Learning", IEEE CIG 2016 ([arXiv:1605.02097](https://arxiv.org/abs/1605.02097))
  Docs: [vizdoom.farama.org](https://vizdoom.farama.org/)

  ### Why This Matters for MOSAIC

  1. **Fills the 3D pixel-based RL gap.** MOSAIC currently emphasises tensor-observation environments (SMAC, MultiGrid, Toy Text) and 2D pixel benchmarks (Atari ALE). ViZDoom fills the vision-based 3D FPS gap where the observation is raw RGB rendered by a game engine with continuous world
  dynamics.

  2. **Directly playable by humans in MOSAIC's Render View.** ViZDoom's rendered window is Doom itself — a game whose controls are well-known and whose gameplay maps naturally onto MOSAIC's keyboard-controlled human worker. Human-vs-RL comparison becomes a first-class experiment (analogous to
   how Google Research Football is used today), which is not natural for tensor-only envs like SMAC.

  3. **Complementary curriculum.** ViZDoom's 8 scenarios form a natural difficulty ramp from `basic` (1 action, 1 target) to `deadly_corridor` (multi-enemy, multi-step). This supports curriculum learning experiments across MOSAIC's operator system: swap out the operator behind an agent
  scenario-by-scenario to measure paradigm-specific difficulty curves.

  4. **Multi-agent routes to heterogeneous team play.** ViZDoom's multiplayer mode supports 2 to 8 agents in deathmatch or cooperative scenarios. This maps directly onto MOSAIC's cross-paradigm mixing vision: a team of one human + one LLM + two RL agents in the same deathmatch match, all
  routed through the Operator abstraction.

  5. **Existing xuance_worker / cleanrl_worker paths cover training.** ViZDoom follows the Gymnasium `env.step(action) → obs, reward, terminated, truncated, info` interface. No new training-worker plumbing required. Just an adapter under `gym_gui/core/adapters/vizdoom.py` following the ALE /
  Crafter pattern.

  ### Proposed Integration

  - **Adapter**: `gym_gui/core/adapters/vizdoom.py`, modelled after `atari_ale.py` (pixel obs, discrete action space) and `crafter.py` (single-agent Gymnasium wrapper)
  - **Enum registration**: add `VIZDOOM` family in `gym_gui/core/enums.py` with the 8 canonical scenarios as `GameId` variants
  - **Docs**: `docs/source/documents/environments/vizdoom/index.rst` with scenario table + demo videos (human + PPO)
  - **Tests**: `gym_gui/tests/test_vizdoom_adapter.py` with smoke tests for load / reset / step / render on `basic.cfg`
  - **Requirements**: `requirements/vizdoom.txt` pinning `vizdoom>=1.2.0`, optional `gymnasium[vizdoom]` extras group

<img width="2180" height="1828" alt="Image" src="https://github.com/user-attachments/assets/c8dc71c8-146e-4ad9-8fb6-367d2c6b35b1" />

  ### References

  Kempka et al., "ViZDoom: A Doom-based AI Research Platform for Visual Reinforcement Learning", IEEE CIG 2016. [arXiv:1605.02097](https://arxiv.org/abs/1605.02097)

**Comment by Abdulhamid97Mousa** [comment]
# ViZDoom is already integrated in MOSAIC, and all 8 canonical scenarios are available through the single-agent training interface.

## play as human with a keyboard

https://github.com/user-attachments/assets/4644af3c-7e87-4ddc-a43b-df0ece99df9f

### What was implemented:
                                                                                                                                                                                                                           
  - **ViZDoomAdapter** wrapping the ViZDoom Gymnasium interface with pixel observations (RGB frames), discrete action space, and auxiliary sensor passthrough (depth buffer, game variables).                              
  - **All 8 scenarios** registered as `env_id` variants: `ViZDoom-Basic-v0`, `ViZDoom-DefendTheCenter-v0`, `ViZDoom-DefendTheLine-v0`, `ViZDoom-HealthGathering-v0`, `ViZDoom-MyWayHome-v0`, `ViZDoom-DeadlyCorridor-v0`,
  `ViZDoom-PredictPosition-v0`, `ViZDoom-TakeCover-v0`.
  - **xuance_worker YAML configs** for ViZDoom covering PPO_Clip and PPG with CNN-based representation.
  - **FastLane video grid** integrated for parallel environment visualization.
  - **Single-agent paradigm form** wired in the MOSAIC GUI.
  - **7/7 smoke tests passing** covering adapter load, reset, step, render, and FastLane frame publish.

**Comment by Abdulhamid97Mousa** [comment]
#  PPO_Clip training on ViZDoom-TakeCover-v0 is fully working. Here is a live session running inside MOSAIC.

https://github.com/user-attachments/assets/586aed3c-45fd-446e-8eaf-c2fd472bca58

  ## Hyperparameters used (reproducible):

  | Parameter | Value |
  |---|---|
  | method | PPO_Clip |
  | env | vizdoom |
  | env_id | ViZDoom-TakeCover-v0 |
  | dl_toolbox | torch |
  | device | cuda:0 |
  | parallels | 4 |
  | running_steps | 5_000_000 |
  | seed | 1 |
  | env_seed | 1 |
  | horizon_size | 256 |
  | learning_rate | 0.0001 |
  | clip_range | 0.2 |
  | ent_coef | 0.01 |
  | gae_lambda | 0.95 |
  | n_epochs | 8 |
  | n_minibatch | 8 |
  | vf_coef | 0.25 |
  | fastlane_enabled | True |
  | fastlane_video_mode | grid |
  | fastlane_grid_limit | 4 |
  | track_tensorboard | True |

  To reproduce, launch `xuance_worker` from the MOSAIC GUI with the above settings. Any of the 8 ViZDoom scenarios can be substituted for `env_id`. For richer pixel visuals, ensure `fastlane_video_mode` is set to
  `grid`.
